### ISSUE TITLE
DoS vulnerability in Revel framework

### ISSUE THREAD

**Issue by SYM01** [revel/revel#1424]
Once the [slices parameter](https://revel.github.io/manual/parameters.html#slices) feature is used, the website will be suffering from DoS attack.

When we need to get a slice parameter, we may use the following code snippet in our controller.
```golang
func (c App) DoS1(name []string) revel.Result {
	return c.RenderText(fmt.Sprint(name))
}

func (c App) DoS2() revel.Result {
	var name []string
	c.Params.Bind(&name, "name")
	return c.RenderText(fmt.Sprint(name))
}
```

It looks like everything is OK. However, we can exhaust the server's MEM with only one request.
e.g., when simply visit `http://localhost:9000/dos1?name[1234567890]=1`, the server's CPU and memory usage will soar, and the OOM killer will be triggered eventually.

This vulnerability was caused by the following code:
https://github.com/revel/revel/blob/a3d7a7c23ca885cc5036d3641ede49ce4a14ee2e/binder.go#L210-L277

When the function above is invoked, Revel will calc the `maxIndex` from user's input, and thus the attacker could make the `maxIndex` as large as possible. When `reflect.MakeSlice` is called, Golang will alloc a large memory as the `maxIndex` needed.

A  possible solution for this vuln is to specify the upper bound of the `maxIndex` in the code or config file.

Before the problem is fixed, plz avoid using the slices parameter feature.

**Comment by notzippy** [comment]
Yes, I can see how that could be an issue, but would it not be a better idea to place the check on where `maxIndex` is set (note the check added on the `if index > -1 && maxIndex < maxIndexBound {` below) ?
```
	maxIndexBound := Config.IntDefault("params.max_index", 1024)

	// Factor out the common slice logic (between form values and files).
	processElement := func(key string, vals []string, files []*multipart.FileHeader) {
		if !strings.HasPrefix(key, name+"[") {
			return
		}

		// Extract the index, and the index where a sub-key starts. (e.g. field[0].subkey)
		index := -1
		leftBracket, rightBracket := len(name), strings.Index(key[len(name):], "]")+len(name)
		if rightBracket > leftBracket+1 {
			index, _ = strconv.Atoi(key[leftBracket+1 : rightBracket])
		}
		subKeyIndex := rightBracket + 1

		// Handle the indexed case.
		if index > -1 && maxIndex < maxIndexBound {
			if index > maxIndex {
				maxIndex = index
			}
			sliceValues = append(sliceValues, sliceValue{
				index: index,
				value: Bind(params, key[:subKeyIndex], typ.Elem()),
			})
			return
		}
...
```

**Comment by SYM01** [comment]
It looks more concise. However, when visiting `https://example.com/?name[1024]=1337`, the developer will still get a slice like `[]string{"1337"}`. It may be confused because it gets `name[0] == "1337"` while the param is `name[1024] == "1337"`.

In my opinion, those illegal parameters should be ignored directly.

**Comment by SYM01** [comment]
Although it looks more concise, it may be confused when visiting `https://example.com/?name[1024]=1`, the developer will get a slice like `[]string{1}` which indicate that `name[0] == "1"`.

**Comment by notzippy** [comment]
OK, so in this situation it would be best to ignore the parameter all together ? And likely a good idea to log the error.
```
		if index > -1  {
                        if maxIndex > maxIndexBound {
                            binderLog.Error.Println("Invalid parameter index, ignoring parameter","index", maxIndex,"key",key)
                            return
                        } 
			if index > maxIndex {
				maxIndex = index
			}
			sliceValues = append(sliceValues, sliceValue{
				index: index,
				value: Bind(params, key[:subKeyIndex], typ.Elem()),
			})
			return
		}
```

**Comment by SYM01** [comment]
It looks better!

**Comment by notzippy** [comment]
Would you mind updating your PR for this bug to match ?

**Comment by brendensoares** [comment]
I've marked this as needs testing, though it may be ready for release. @notzippy any idea what the latest status is here?

**Comment by brendensoares** [comment]
pinging this; high priority

**Comment by notzippy** [comment]
@brendensoares This has been fixed and merged see #1427 and should be closed 

https://github.com/revel/revel/blob/master/binder.go#L232-L247
